File: ios/Potillus/AppOverflowMenu.swift
// vim: set et ts=4:
// =============================================================================
// Libellus Potionis - Privacy-Friendly Alcohol Tracker
// Copyright (c) 2026 Martin A. Godisch <martin@godisch.de>
// =============================================================================
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program.  If not, see <https://www.gnu.org/licenses/>.
//
// In addition, as permitted by section 7 of the GNU General Public License,
// this program may carry additional permissions; any such permissions that
// apply to it are stated in the accompanying COPYING.md file.
//
// =============================================================================

import PotillusKit
import SwiftUI

// =============================================================================
// AppOverflowMenu – the toolbar menu shared by all four main screens
// =============================================================================
//
// Android carries one `AppOverflowMenu` composable in the top bar of every main
// screen, so the menu's entries live in exactly one place instead of being
// copied four times. This is the iOS twin: a single `ViewModifier` that each of
// Today / Calendar / Statistics / Drinks applies with `.appOverflowMenu(...)`.
//
// WHY A MODIFIER, NOT A PLAIN VIEW
//   The menu is not just a button — it owns the sheets it opens (Settings and
//   the Copyright viewer) and the state that drives them. A `ViewModifier` lets
//   all of that ride along on the screen it decorates, so a screen adds the menu,
//   its two destinations and their presentation state in one line.
//
// PRESENTATION, NOT PORT
//   The entries match Android (Settings, Help, Lock app, About). The FORM is
//   native: a SwiftUI `Menu` behind the ellipsis "More" affordance at the
//   trailing edge of the navigation bar — where Photos, Files and Notes put
//   theirs — not a Material dropdown. Same choices, native idiom — the rule the
//   rest of this port follows. (Until the 0.84.0 QA round the menu wore
//   Android's hamburger at the LEADING edge; iOS has no hamburger convention,
//   so it was the one Android-ism left in the navigation.)
// =============================================================================

struct AppOverflowMenu: ViewModifier {

    /// Needed to build the Settings sheet; the value struct is passed in rather
    /// than read from the environment because `AppEnvironment` is a plain
    /// `Sendable` struct, not an `@Observable` the environment can vend by type.
    let environment: AppEnvironment

    @Environment(\.appLocale) private var locale

    /// The lock is an `@Observable` injected once by `RootView`, so it is read
    /// from the environment rather than threaded through four screen initialisers.
    /// Optional so a screen previewed without the injection renders instead of
    /// trapping; the "Lock app" entry simply does not appear.
    @Environment(AppLockModel.self) private var lock: AppLockModel?

    @State private var showingSettings = false
    @State private var showingHelp = false
    @State private var showingAbout = false

    func body(content: Content) -> some View {
        content
            .toolbar {
                // Trailing, as iOS's own apps place their More menus. The
                // screen's primary action ("+") and its edit toggle declare their
                // own trailing items; SwiftUI orders items from separate toolbar
                // blocks itself, so the exact neighbour order is the platform's
                // call — confirm it looks right on a device after changes here.
                ToolbarItem(placement: .topBarTrailing) {
                    Menu {
                        Button {
                            showingSettings = true
                        } label: {
                            Label(Loc.string("Settings", locale: locale), systemImage: "gearshape")
                        }
                        .accessibilityIdentifier("nav.settings")
                        Button {
                            showingHelp = true
                        } label: {
                            Label(Loc.string("Help", locale: locale), systemImage: "questionmark.circle")
                        }
                        // "Lock app" appears whenever the device can authenticate,
                        // matching Android: a manual lock no longer requires auto-lock
                        // to be armed, and lockNow()/retry() can always clear the cover
                        // again. Hidden only when there is no biometric and no passcode,
                        // where a cover would strand the user.
                        if lock?.deviceCanAuthenticate() == true {
                            Button {
                                Task { await lock?.lockNow() }
                            } label: {
                                Label(Loc.string("Lock app", locale: locale), systemImage: "lock")
                            }
                        }
                        // About LAST, after the conditional "Lock app", as on
                        // Android: version and licenses are looked up once, not
                        // daily, so the entry yields the prime positions to the
                        // three that do real work.
                        Button {
                            showingAbout = true
                        } label: {
                            Label(Loc.string("About", locale: locale), systemImage: "info.circle")
                        }
                    } label: {
                        // ellipsis.circle is the platform's "More" glyph; the
                        // label keeps the catalogue's translated "Menu", which
                        // says what opens without needing a 22nd translation run.
                        Label(Loc.string("Menu", locale: locale), systemImage: "ellipsis.circle")
                    }
                    .accessibilityIdentifier("nav.menu")
                }
            }
            .sheet(isPresented: $showingSettings) {
                SettingsScreen(environment: environment)
            }
            .sheet(isPresented: $showingHelp) {
                // The user guide, in the app's language with an English fallback.
                NavigationStack {
                    DocumentViewerScreen(
                        title: Loc.string("Help", locale: locale),
                        resource: guideResource()
                    )
                    .toolbar {
                        ToolbarItem(placement: .confirmationAction) {
                            Button(Loc.string("Done", locale: locale)) {
                                showingHelp = false
                            }
                        }
                    }
                }
            }
            .sheet(isPresented: $showingAbout) {
                // About is now reached from the overflow menu (as on Android),
                // not from Settings. It carries its own stack so its "Copyright &
                // license" link can push the full document, and a Done button to
                // dismiss the sheet.
                NavigationStack {
                    AboutScreen()
                        .toolbar {
                            ToolbarItem(placement: .confirmationAction) {
                                Button(Loc.string("Done", locale: locale)) {
                                    showingAbout = false
                                }
                            }
                        }
                }
            }
    }

    /// The bundled guide for the app's language, English as the guaranteed
    /// fallback. Tries the exact tag (`usersguide_zh-Hant`), then the base
    /// language (`usersguide_de` for a `de-DE` system locale), then
    /// `usersguide_en`. Only the guides whose templates have been authored ship,
    /// so an as-yet-untranslated language resolves to English rather than a blank
    /// page.
    private func guideResource() -> String {
        var candidates: [String] = []
        let tag = locale.identifier(.bcp47)
        if !tag.isEmpty {
            candidates.append("usersguide_\(tag)")
            if let dash = tag.firstIndex(of: "-") {
                candidates.append("usersguide_\(tag[..<dash])")
            }
        }
        candidates.append("usersguide_en")
        for name in candidates where Bundle.main.url(forResource: name, withExtension: "md") != nil {
            return name
        }
        return "usersguide_en"
    }
}

extension View {
    /// Add the shared overflow menu (Settings, Help, Copyright, Lock app) to a
    /// screen's navigation bar. Apply it inside the screen's `NavigationStack`, on
    /// the same view that carries `.navigationTitle`, so the button lands in that
    /// bar.
    func appOverflowMenu(environment: AppEnvironment) -> some View {
        modifier(AppOverflowMenu(environment: environment))
    }
}

File: android/app/src/main/kotlin/de/godisch/potillus/ui/component/AppOverflowMenu.kt
/* vim: set et ts=4:
 * =============================================================================
 * Libellus Potionis - Privacy-Friendly Alcohol Tracker
 * Copyright (c) 2026 Martin A. Godisch <martin@godisch.de>
 * =============================================================================
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * In addition, as permitted by section 7 of the GNU General Public License,
 * this program may carry additional permissions; any such permissions that
 * apply to it are stated in the accompanying COPYING.md file.
 *
 * =============================================================================
 */
package de.godisch.potillus.ui.component

import androidx.biometric.BiometricManager
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG
import androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Help
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.Menu
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import de.godisch.potillus.R

/**
 * The top-bar overflow menu shared by all four main screens.
 *
 * WHY A SINGLE SHARED COMPOSABLE?
 *   Previously each of the four screens (Today / Calendar / Statistics / Drinks)
 *   carried an identical settings `IconButton` in its `TopAppBar`. That was four
 *   copies of the same UI to keep in sync. This composable replaces all of them
 *   with one definition, so the menu's icon, entries and order live in exactly
 *   one place.
 *
 * BEHAVIOUR
 *   The button shows a "burger" icon ([Icons.Default.Menu]). Tapping it opens a
 *   [DropdownMenu] (an Android "overflow menu") anchored to the button, with
 *   four entries in this order: Settings, Help, "Lock app" and About. Selecting
 *   an entry closes the menu and invokes the matching callback; the actual
 *   navigation is the caller's responsibility (see
 *   [de.godisch.potillus.ui.nav.AppNavigation]), which keeps this component free
 *   of any navigation dependency.
 *
 * WHY ABOUT COMES LAST
 *   It is the entry a user reaches for least often -- version and licences are
 *   looked up once, not daily -- so it yields the prime positions to the three
 *   entries that do real work. iOS orders the same menu identically.
 *
 * WHY THESE GLYPHS
 *   A question mark in a circle for Help and an "i" in a circle for About are
 *   the conventional pair, and the same metaphors iOS uses (`questionmark.circle`
 *   and `info.circle`). The FILL differs by platform on purpose: this menu's
 *   other entries are filled glyphs, so an outlined circle between them would
 *   read as a different weight class, whereas on iOS the outlined SF Symbols are
 *   what sits naturally beside `gearshape` and `lock`. Metaphor is shared; fill
 *   follows each platform's own convention.
 *
 *   Note [Icons.AutoMirrored.Filled.Help], not `Icons.Filled.Help`: the latter
 *   is deprecated in favour of the auto-mirrored set, because a question mark
 *   mirrors in right-to-left layouts. [Icons.Filled.Info] has no auto-mirrored
 *   variant -- an "i" in a circle looks the same either way -- and lives in
 *   material-icons-core rather than -extended.
 *
 * @param onOpenSettings Invoked when the "Settings" entry is chosen.
 * @param onOpenHelp     Invoked when the "Help" entry is chosen (opens the
 *                       in-app user guide viewer).
 * @param onOpenAbout    Invoked when the "About" entry is chosen (opens the
 *                       About screen: version, licence and components).
 * @param onLockApp      Invoked when the "Lock app" entry is chosen (manually locks
 *                       the app). The entry is only shown when an authenticator
 *                       (biometric or device credential) is available — otherwise
 *                       locking would strand the user (Variant A).
 * @param tint           Colour for the burger icon. Defaults to the ambient
 *                       content colour; the main screens pass their top-bar
 *                       `onPrimary` colour so the icon matches the former gear.
 */
@Composable
fun AppOverflowMenu(
    onOpenSettings: () -> Unit,
    onOpenHelp: () -> Unit,
    onOpenAbout: () -> Unit,
    onLockApp: () -> Unit,
    tint: Color = LocalContentColor.current,
) {
    // `expanded` is the only piece of state this component owns: whether the
    // dropdown is currently shown. `remember` keeps it across recompositions;
    // it intentionally does NOT survive process death, which is correct for a
    // transient menu.
    var expanded by remember { mutableStateOf(false) }

    // Variant A: the "Lock app" entry is shown only when the device can actually
    // authenticate (strong biometric OR device credential). Without an
    // authenticator, manually locking would leave no way back in, so the entry is
    // hidden. Computed once per composition instance (cheap binder call, cached).
    val context = LocalContext.current
    val canLock = remember {
        BiometricManager.from(context)
            .canAuthenticate(BIOMETRIC_STRONG or DEVICE_CREDENTIAL) ==
            BiometricManager.BIOMETRIC_SUCCESS
    }

    IconButton(onClick = { expanded = true }) {
        Icon(
            imageVector = Icons.Default.Menu,
            contentDescription = stringResource(R.string.menu),
            tint = tint,
        )
    }

    DropdownMenu(
        expanded = expanded,
        onDismissRequest = { expanded = false },
    ) {
        // Each item closes the menu BEFORE invoking its callback so the dropdown
        // is already dismissed by the time the destination is pushed.
        DropdownMenuItem(
            text = { Text(stringResource(R.string.settings)) },
            leadingIcon = { Icon(Icons.Filled.Settings, contentDescription = null) },
            onClick = {
                expanded = false
                onOpenSettings()
            },
        )
        DropdownMenuItem(
            text = { Text(stringResource(R.string.help)) },
            // A question mark in a circle. No explicit `tint` is set, so the icon
            // inherits the menu's ambient content colour and blends in with the
            // theme.
            leadingIcon = { Icon(Icons.AutoMirrored.Filled.Help, contentDescription = null) },
            onClick = {
                expanded = false
                onOpenHelp()
            },
        )
        // "Lock app" — manual lock (Variant A). Only present when an authenticator
        // is available, so it never leaves the user unable to get back in.
        if (canLock) {
            DropdownMenuItem(
                text = { Text(stringResource(R.string.lock_app)) },
                leadingIcon = { Icon(Icons.Filled.Lock, contentDescription = null) },
                onClick = {
                    expanded = false
                    onLockApp()
                },
            )
        }
        // About LAST, after the conditional "Lock app": on a device with no
        // authenticator the menu simply closes up around the gap, and About is
        // the final entry either way.
        DropdownMenuItem(
            text = { Text(stringResource(R.string.about)) },
            // An "i" in a circle — the conventional glyph for "information about
            // this app", and the metaphor iOS uses with `info.circle`.
            leadingIcon = { Icon(Icons.Filled.Info, contentDescription = null) },
            onClick = {
                expanded = false
                onOpenAbout()
            },
        )
    }
}
